Deprecate replacing full context object of `EmbeddedViewRef`

devversion opened the pull request:
See individual commits

devversion committed:
test: add benchmark for ng template outlet context modifications

Adds a benchmark for potential upcoming `NgTemplateOutlet` context
logic updates.

devversion committed:
build: fix IDE completion for benchmarks code

We are including a file, for some reason, that is outside of the root
dir. This breaks TypeScript/ or more specifically VSCode from picking
up this tsconfig- breaking path mappings and hence auto completion.

devversion commented on packages/common/src/directives/ng_template_outlet.ts:
```diff
@@ -57,30 +57,47 @@ export class NgTemplateOutlet<C = unknown> implements OnChanges {
+      // For a given outlet instance, we create a proxy object that delegates
+      // to the user-specified context. This allows changing, or swapping out
+      // the context object completely without having to destroy/re-create the view.
+      const proxyContext = <C>new Proxy({}, {
```
do we know if this is something worth optimizing? I'm trying to do the change, but it seems to add additional complexity that I'm not sure is worth doing (plus the need to re-create the view- although I agree that the common case is with context=undefined)

devversion replied:
Pushed a change for this (went with a boolean flag to avoid checking the view ref context object)

devversion replied:
Updated based on what we discussed. i.e. we are not going with the optimization

crisbeto reviewed (approved):
LGTM

devversion committed:
refactor(core): deprecate allowing full context object to be replaced in `EmbeddedViewRef`

This partially reverts commit a3e17190e7e7e0329ed3643299c24d5fd510b7d6
and deprecates behavior added.

The context of an embedded view ref at some point was switched from a
getter to an actual assignable property. This is something we revert
as it introduces additional complexity for our generated code
(in terms of closures capturing the `ctx`), creates technical
limitations for Angular's internals and the usage pattern is rarely
used (and can be addressed via simple assignments, `Object.assign` or
the use of a proxy if replacing the full context object is still
desirable)

DEPRECATED: Swapping out the context object for `EmbeddedViewRef`
is no longer supported. Support for this was introduced with v12.0.0, but
this pattern is rarely used. There is no replacement, but you can use
simple assignments in most cases, or `Object.assign , or alternatively
still replace the full object by using a `Proxy` (see `NgTemplateOutlet`
as an example).

Also adds a warning if the deprecated

devversion commented:
[TGP](https://fusion2.corp.google.com/presubmit/568131797/OCL:568131797:BASE:569143990:1695928035436:1ad746c5/targets)

pkozlowski-opensource commented on packages/common/src/directives/ng_template_outlet.ts:
```diff
@@ -36,48 +36,35 @@ import {Directive, EmbeddedViewRef, Injector, Input, OnChanges, SimpleChanges, T
-  /** @nodoc */
   ngOnChanges(changes: SimpleChanges) {
-    if (changes['ngTemplateOutlet'] || changes['ngTemplateOutletInjector']) {
+    if (changes['ngTemplateOutlet']) {
```
This looks like a change in the behaviour where we drop embedded view re-creation on the injector change. Is this intentional?

pkozlowski-opensource replied:
Oh, I can see that it is changed again in the subsequent commit, but maybe we shouldn't have those changes here then?

devversion replied:
Yeah, the idea was to revert Kristiyan's commit in 3rd and then in 4th switch `NgTemplateOutlet` to not rely on the deprecated property. I think that was messed up. Will try to fix but skip another TGP.

pkozlowski-opensource commented on packages/common/src/directives/ng_template_outlet.ts:
```diff
@@ -36,48 +36,35 @@ import {Directive, EmbeddedViewRef, Injector, Input, OnChanges, SimpleChanges, T
-export class NgTemplateOutlet<C = unknown> implements OnChanges {
-  private _viewRef: EmbeddedViewRef<C>|null = null;
+export class NgTemplateOutlet implements OnChanges {
+  private _viewRef: EmbeddedViewRef<any>|null = null;
```
Wonder about this change. Shouldn't we use `Object` to align with the type of ngTemplateOutletContext?

devversion replied:
(this was just visible on the revert commit - resolving)

pkozlowski-opensource reviewed (approved):
LGTM. I think I was confused about the sequence of changes between the 4 commits. The end version looks good, but the 3rd commit was a bit surprising.

pkozlowski-opensource reviewed (approved):
LGTM

Reviewed-for: public-api 
Reviewed-for: size-tracking
Reviewed-for: fw-core 
Reviewed-for: fw-common

devversion committed:
refactor(common): update `NgTemplateOutlet` to no longer rely on context swapping

The context of an embedded view ref at some point was switched from a
getter to an actual assignable property. This is something we reverted
with the previous commit as it introduces additional complexity for our
generated code (in terms of closures capturing the `ctx`).

This change impacted the template outlet code because we actively relied
on swapping out the full context if the user changes it. Previousl,
before we allowed to swap out the context (in v16), we mutated the
initial view context if it didn't change structurally- and in other
cases the view was re-created. We improved this performance aspect with
the changes to allow for the context to be swapped out + actually also
fixed a bug where the initial context object was mutated and the user
could observe this change.

This commit adjusts for context not being replacable- while still
keeping the bugs fixed and preserving the performance wins of not
having to destroy/re-create the view whenever the context changes.

Benchmarks: https://hackmd.io/J0Ci_JzxQ0K1AA1omXhIQQ

devversion commented:
@pkozlowski-opensource fixed the commit confusion.

Caretaker note: Please ignore the `saucelabs` check. These are unrelated failures.

alxhub commented:
This PR was merged into the repository by commit 7426948ff63a68524cee19f891205525bb9d5bd7.
